mmedit/models/inpaintors/aot_inpaintor.py

```diff
@@ -0,0 +1,279 @@
+            is_real (bool): If True, the gan loss will regard this batch as
+                real data. Otherwise, the gan loss will regard this batch as
+                fake data.
+            is_disc (bool): If True, this function is called in discriminator
+                training step. Otherwise, this function is called in generator
+                training step. This will help us to compute different types of
+                adversarial loss, like LSGAN.
+            **kwargs (keyword arguments).
+        Returns:
+            dict: Contains the loss items computed in this function.
+        """
+
+        pred = self.disc(data_batch)
+        loss_ = self.loss_gan(pred, is_real, is_disc, **kwargs)
+
+        loss = dict(real_loss=loss_) if is_real else dict(fake_loss=loss_)
+
+        if self.with_disc_shift_loss:
+            loss_d_shift = self.loss_disc_shift(loss_)
+            # 0.5 for average the fake and real data
+            loss.update(loss_disc_shift=loss_d_shift * 0.5)
+
+        return loss
+
+    def generator_loss(self, fake_res, fake_img, data_batch):
+        """Forward function in generator training step.
+        In this function, we mainly compute the loss items for generator with
+        the given (fake_res, fake_img). In general, the `fake_res` is the
+        direct output of the generator and the `fake_img` is the composition of
+        direct output and ground-truth image.
+        Args:
+            fake_res (torch.Tensor): Direct output of the generator.
+            fake_img (torch.Tensor): Composition of `fake_res` and
+                ground-truth image.
+            data_batch (dict): Contain other elements for computing losses.
+        Returns:
+            tuple(dict): Dict contains the results computed within this \
+                function for visualization and dict contains the loss items \
+                computed in this function.
+        """
+        gt = data_batch['gt_img']
+        mask = data_batch['mask']
+        masked_img = data_batch['masked_img']
+
+        loss = dict()
+
+        if self.with_gan:
+            pred = self.disc(fake_img)
+            loss_g_fake = self.loss_gan(pred, True, False, mask=mask)
+            loss['loss_g_fake'] = loss_g_fake
+
+        if self.with_l1_valid_loss:
+            loss_loss_l1_valid = self.loss_l1_valid(fake_res, gt)
+            loss['loss_l1_valid'] = loss_loss_l1_valid
+
+        if self.with_out_percep_loss:
+            loss_out_percep, loss_out_style = self.loss_percep(fake_res, gt)
+            if loss_out_percep is not None:
+                loss['loss_out_percep'] = loss_out_percep
+            if loss_out_style is not None:
+                loss['loss_out_style'] = loss_out_style
+
+        res = dict(
+            gt_img=gt.cpu(),
+            masked_img=masked_img.cpu(),
+            fake_res=fake_res.cpu(),
+            fake_img=fake_img.cpu())
+
+        return res, loss
+
+    def forward_test(self,
```

Comment:
This should be identical to that in `OneStageInpaintor`?

Reply:
again there are a few subtle differences. For example, the holes of `masked_img` for aot inpaintor are filled with ones by combining with mask:

![image](https://user-images.githubusercontent.com/33197366/149704646-917e29b6-1dea-4f5d-9b8c-e6c05524b607.png)

whereas `masked_img` for onestageinpaintor has holes filled with zeros.

Another difference is the `data_dict` used for calculating the metrics:

For aot inpaintor:

![image](https://user-images.githubusercontent.com/33197366/149704924-b7fb92ee-e366-4149-9884-40fb801f7de9.png)

For onestageinpaintor:

![image](https://user-images.githubusercontent.com/33197366/149704990-86e0234a-f9ae-4988-8b12-d1a750348d32.png)